File: core/src/com/mygdx/game/MyGdxGame.java
package com.mygdx.game;

import com.badlogic.ashley.core.Engine;
import com.badlogic.ashley.core.Entity;
import com.badlogic.gdx.ApplicationAdapter;
import com.badlogic.gdx.Gdx;
import com.badlogic.gdx.Input;
import com.badlogic.gdx.graphics.GL20;
import com.badlogic.gdx.graphics.OrthographicCamera;
import com.badlogic.gdx.graphics.g2d.Animation;
import com.badlogic.gdx.graphics.g2d.BitmapFont;
import com.badlogic.gdx.graphics.g2d.SpriteBatch;
import com.badlogic.gdx.graphics.g2d.TextureAtlas;
import com.badlogic.gdx.graphics.g2d.TextureRegion;
import com.badlogic.gdx.math.Vector2;
import com.badlogic.gdx.utils.Array;

public class MyGdxGame extends ApplicationAdapter {
  private static final float PLAYER_SIZE = 100f;
  SpriteBatch batch;
  SpriteBatch hudBatch;
  TextureRegion img;
  TextureRegion box;
  TextureAtlas atlas;
  Animation<TextureRegion> appleAnimation;
  Animation<TextureRegion> runningAnimation;
  Vector2 playerPosition = new Vector2();
  float speedX = 200f;
  float speedY = 200f;
  boolean running;
  float runningTime = 0;
  OrthographicCamera camera;
  BitmapFont font;

  Engine engine;

  @Override
  public void create() {
    batch = new SpriteBatch();
    hudBatch = new SpriteBatch();
    atlas = new TextureAtlas("packed/assets.atlas");
    box = atlas.findRegion("box");
    img = atlas.findRegion("idle");
    Array<TextureAtlas.AtlasRegion> appleFrames = atlas.findRegions("apple");
    appleFrames.removeRange(0, 4);
    appleAnimation = new Animation<TextureRegion>(0.05f, appleFrames, Animation.PlayMode.LOOP);
    runningAnimation = new Animation<TextureRegion>(0.05f, atlas.findRegions("run"), Animation.PlayMode.LOOP);
    font = new BitmapFont(Gdx.files.internal("font/font.fnt"));

    camera = new OrthographicCamera(Gdx.graphics.getWidth(), Gdx.graphics.getHeight());

    engine = new Engine();
    engine.addSystem(new AnimationSystem());
    engine.addSystem(new AnimationRenderingSystem(batch));

    Entity apple = new Entity();
    apple.add(new AnimationComponent(appleAnimation));
    apple.add(new PositionComponent(300f, 350f));
    engine.addEntity(apple);

  }

  @Override
  public void render() {
    float time = Gdx.graphics.getDeltaTime();
    Vector2 move = new Vector2();
    if (Gdx.input.isKeyPressed(Input.Keys.A)) {
      move.x = -speedX * time;
    } else if (Gdx.input.isKeyPressed(Input.Keys.D)) {
      move.x = speedX * time;
    }

    if (Gdx.input.isKeyPressed(Input.Keys.W)) {
      move.y = speedY * time;
    } else if (Gdx.input.isKeyPressed(Input.Keys.S)) {
      move.y = -speedY * time;
    }

    float cameraScale = Gdx.input.isKeyPressed(Input.Keys.SPACE) ? 2f : 1f;
    camera.viewportWidth = Gdx.graphics.getWidth() * cameraScale;
    camera.viewportHeight = Gdx.graphics.getHeight() * cameraScale;

    running = !move.isZero();
    playerPosition.add(move);
    if (running) {
      runningTime += time;
    } else {
      runningTime = 0;
    }
    camera.position.x = playerPosition.x;
    camera.position.y = playerPosition.y + 200;

    Gdx.gl.glClearColor(0.9f, 0.9f, 0.9f, 1);
    Gdx.gl.glClear(GL20.GL_COLOR_BUFFER_BIT);

    camera.update();
    batch.setProjectionMatrix(camera.combined);

    engine.update(time);

    batch.begin();
    if (running) {
      TextureRegion frame = runningAnimation.getKeyFrame(runningTime);
      batch.draw(frame, playerPosition.x, playerPosition.y,
          PLAYER_SIZE / 2, PLAYER_SIZE / 2,
          PLAYER_SIZE, PLAYER_SIZE,
          move.x >= 0 ? 1 : -1, 1,
          0
      );
    } else {
      batch.draw(img, playerPosition.x, playerPosition.y, 100f, 100f);
    }

    batch.draw(box, 500, 450, 100f, 100f);
    batch.draw(box, 0, 250, 70f, 70f);
    batch.draw(box, 650, -30, 130f, 130f);

    batch.end();

    hudBatch.begin();
    font.draw(hudBatch, "Hello, world!", 100, 100);
    hudBatch.end();
  }

  @Override
  public void dispose() {
    batch.dispose();
    atlas.dispose();
  }
}

File: core/src/com/mygdx/game/MyGdxGame.kt
package com.mygdx.game

import com.badlogic.ashley.core.Engine
import com.badlogic.ashley.core.Entity
import com.badlogic.gdx.ApplicationAdapter
import com.badlogic.gdx.Gdx
import com.badlogic.gdx.graphics.GL20
import com.badlogic.gdx.graphics.OrthographicCamera
import com.badlogic.gdx.graphics.g2d.Animation
import com.badlogic.gdx.graphics.g2d.BitmapFont
import com.badlogic.gdx.graphics.g2d.SpriteBatch
import com.badlogic.gdx.graphics.g2d.TextureAtlas
import com.mygdx.game.components.*
import com.mygdx.game.systems.*

class MyGdxGame : ApplicationAdapter() {
  lateinit var batch: SpriteBatch
  lateinit var hudBatch: SpriteBatch
  lateinit var atlas: TextureAtlas
  lateinit var camera: OrthographicCamera
  lateinit var font: BitmapFont
  lateinit var engine: Engine

  lateinit var playerPositionComponent: PositionComponent

  override fun create() {
    batch = SpriteBatch()
    hudBatch = SpriteBatch()
    atlas = TextureAtlas("packed/assets.atlas")
    font = BitmapFont(Gdx.files.internal("font/font.fnt"))

    camera = OrthographicCamera(Gdx.graphics.width.toFloat(), Gdx.graphics.height.toFloat())

    val idleAnimation = Animation(0.05f, atlas.findRegions("idle"), Animation.PlayMode.LOOP)
    val runAnimation = Animation(0.05f, atlas.findRegions("run"), Animation.PlayMode.LOOP)

    engine = Engine()
    engine.addSystem(InputSystem())
    engine.addSystem(PlayerStateSystem(idleAnimation, runAnimation))
    engine.addSystem(MovementSystem())
    engine.addSystem(FlippingSystem())
    engine.addSystem(AnimationSystem())
    engine.addSystem(CameraSystem(camera, batch))
    engine.addSystem(RenderingSystem(batch))
    engine.addSystem(HudRenderingSystem(hudBatch, font))

    engine.addEntity(appleEntity())

    engine.addEntity(boxEntity(500f, 450f, 100f))
    engine.addEntity(boxEntity(0f, 250f, 70f))
    engine.addEntity(boxEntity(650f, -30f, 130f))

    engine.addEntity(Entity().apply {
      add(PositionComponent(100f, 100f))
      add(TextComponent("hello, world!"))
    })

    playerPositionComponent = PositionComponent(0f, 0f)
    engine.addEntity(Entity().apply {
      add(playerPositionComponent)
      add(SizeComponent(100f, 100f))
      add(AnimationComponent(idleAnimation))
      add(TextureComponent(idleAnimation.getKeyFrame(0f)))
      add(MoveComponent())
      add(ScaleComponent())
      add(PlayerStateComponent())
    })
  }

  private fun appleEntity(): Entity =
      Entity().apply {
        val animationComponent = AnimationComponent(Animation(
            0.05f,
            atlas.findRegions("apple").apply { removeRange(0, 4) },
            Animation.PlayMode.LOOP))
        add(animationComponent)
        add(TextureComponent(animationComponent.currentFrame))
        add(PositionComponent(300f, 350f))
        add(SizeComponent(75f, 75f))
      }

  private fun boxEntity(x: Float, y: Float, size: Float) =
      Entity().apply {
        add(TextureComponent(atlas.findRegion("box")))
        add(PositionComponent(x, y))
        add(SizeComponent(size, size))
      }

  override fun render() {
    Gdx.gl.glClearColor(0.9f, 0.9f, 0.9f, 1f)
    Gdx.gl.glClear(GL20.GL_COLOR_BUFFER_BIT)
    engine.update(Gdx.graphics.deltaTime)
  }

  override fun dispose() {
    batch.dispose()
    atlas.dispose()
  }
}